Src/Common/CorrelationIdLookupHelper.cs

```diff
@@ -0,0 +1,158 @@
+﻿namespace Microsoft.ApplicationInsights.Common
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Globalization;
+    using System.IO;
+    using System.Net;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// A store for instrumentation App Ids. This makes sure we don't query the public endpoint to find an app Id for the same Ikey more than once.
+    /// </summary>
+    internal static class CorrelationIdLookupHelper
+    {
+        /// <summary>
+        /// Max number of app ids to cache.
+        /// </summary>
+        private const int MAXSIZE = 100;
+
+        private const int GET_APP_ID_TIMEOUT = 2000; // milliseconds
+
+        private const string CORRELATION_ID_FORMAT = "aid-v1:{0}";
+
+        private const string APPID_QUERY_API_RELATIVE_URI_FORMAT = "api/profiles/{0}/appId";
+
+        private static ConcurrentDictionary<string, string> knownCorrelationIds = new ConcurrentDictionary<string, string>();
```

Comment:
provide a configuration experience to pre-populate this dictionary by setting property on corresponding telemetry module. If customer wants to run it once in startup task or during the build.